lib/locale_threads.t

```diff
@@ -54,67 +471,1305 @@ SKIP: { # perl #127708
+    # that locale, then check that the result matches the expected value, and
+    # to immediately repeat some largish number of iterations.  The goal is to
+    # see if the locales on each thread are truly independent of those on the
+    # other threads.
+    #
+    # To that end, the locales are chosen so that the results differ from
+    # every other locale.  Otherwise, the thread results wouldn't be truly
+    # independent.  But if there are more threads than there are distinct
+    # results, duplicates are used to fill up what would otherwise be empty
+    # slots.  That is the best we can do on those platforms.
+    #
+    # Having lots of locales to continually switch between stresses things so
+    # as to find potential segfaults where locale changing isn't really thread
+    # safe.
 
-        my \@threads = map +threads->create(sub {
-            sleep 0.1;
-            for (1..5_000) {
-                my \$s;
-                my \$in = 4.2; # avoid any constant folding bugs
-
-                if ('$comma') {
-                    setlocale(&LC_NUMERIC, '$comma');
-                    use locale;
-                    \$s = sprintf('%g', \$in);
-                    return 0 if (\$s ne '4,2');
+    # There is a bug in older Windows runtimes in which locales in CP1252 and
+    # similar code pages whose names aren't entirely ASCII aren't recognized
+    # by later setlocales.  Some names that are all ASCII are synonyms for
+    # such names.  Weed those out by doing a setlocale of the original name,
+    # and then a setlocale of the resulting one.  Discard locales which have
+    # any unacceptable name
+    if (${^O} eq "MSWin32" && $Config{'libc'} !~ /ucrt/) {
+        for my $i (0 .. $#locales) {
+            my $locale_name = $locales[$i]->{locale_name};
+            my $underlying_name = setlocale(&LC_CTYPE, $locale_name);
+
+            # Defeat any attempt to skip the setlocale if the same as current,
+            # by switching to a locale very unlikey to be the current one.
+            setlocale($LC_ALL, "Albanian");
+
+            next if setlocale(&LC_CTYPE, $underlying_name);
+
+            splice @locales, $i, 1;
+        }
+    }
+
+    # Create a hash of the errnos:
+    #          "1" => "Operation\\ not\\ permitted",
+    #          "2" => "No\\ such\\ file\\ or\\ directory",
+    #          etc.
+    my %msg_catalog;
+    foreach my $error (sort keys %!) {
+        my $number = eval "Errno::$error";
+        $! = $number;
+        my $description = "$!";
+        next unless "$description";
+        $msg_catalog{$number} = quotemeta "$description";
+    }
+
+    # Then just the errnos.
+    my @msg_catalog = sort { $a <=> $b } keys %msg_catalog;
+
+    # Remove the excess ones.
+    splice @msg_catalog, $max_message_catalog_entries
+                                          if $max_message_catalog_entries >= 0;
+    my $msg_catalog = join ',', @msg_catalog;
+
+    eval "POSIX::localeconv()->{currency_symbol}";
```

Comment:
Why is this a string eval?  Wouldn't:
```
eval  { POSIX::localeconv()->{currency_symbol}; };
```
be equivalent?

Reply:
Yes, thanks; changed